bridge/prometheus/producer.go

```diff
@@ -0,0 +1,253 @@
+	for _, gatherer := range p.gatherers {
+		promMetrics, err := gatherer.Gather()
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		m, err := convertPrometheusMetricsInto(promMetrics, now)
+		otelMetrics = append(otelMetrics, m...)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if errs.errOrNil() != nil {
+		otel.Handle(errs.errOrNil())
+	}
+	if len(otelMetrics) == 0 {
+		return nil, nil
+	}
+	return []metricdata.ScopeMetrics{{
+		Scope: instrumentation.Scope{
+			Name: scopeName,
+		},
+		Metrics: otelMetrics,
+	}}, nil
+}
+
+func convertPrometheusMetricsInto(promMetrics []*dto.MetricFamily, now time.Time) ([]metricdata.Metrics, error) {
+	var errs multierr
+	otelMetrics := make([]metricdata.Metrics, 0)
+	for _, pm := range promMetrics {
+		newMetric := metricdata.Metrics{
+			Name:        pm.GetName(),
+			Description: pm.GetHelp(),
+		}
+		switch pm.GetType() {
+		case dto.MetricType_GAUGE:
+			newMetric.Data = convertGauge(pm.GetMetric(), now)
+		case dto.MetricType_COUNTER:
+			newMetric.Data = convertCounter(pm.GetMetric(), now)
+		case dto.MetricType_HISTOGRAM:
+			newMetric.Data = convertHistogram(pm.GetMetric(), now)
+		default:
+			// MetricType_GAUGE_HISTOGRAM, MetricType_SUMMARY, MetricType_UNTYPED
+			errs = append(errs, fmt.Errorf("%w: %v for metric %v", errUnsupportedType, pm.GetType(), pm.GetName()))
+			continue
+		}
+		otelMetrics = append(otelMetrics, newMetric)
+	}
+	return otelMetrics, errs.errOrNil()
+}
+
+func convertGauge(metrics []*dto.Metric, now time.Time) metricdata.Gauge[float64] {
+	otelGauge := metricdata.Gauge[float64]{
+		DataPoints: make([]metricdata.DataPoint[float64], len(metrics)),
+	}
+	for i, m := range metrics {
+		dp := metricdata.DataPoint[float64]{
+			Attributes: convertLabels(m.GetLabel()),
+			Time:       now,
+			Value:      m.GetGauge().GetValue(),
+		}
+		if m.GetTimestampMs() != 0 {
+			dp.Time = time.UnixMilli(m.GetTimestampMs())
+		}
+		otelGauge.DataPoints[i] = dp
+	}
+	return otelGauge
+}
+
+func convertCounter(metrics []*dto.Metric, now time.Time) metricdata.Sum[float64] {
+	otelCounter := metricdata.Sum[float64]{
+		DataPoints:  make([]metricdata.DataPoint[float64], len(metrics)),
+		Temporality: metricdata.CumulativeTemporality,
+		IsMonotonic: true,
+	}
+	for i, m := range metrics {
+		dp := metricdata.DataPoint[float64]{
+			Attributes: convertLabels(m.GetLabel()),
+			StartTime:  processStartTime,
+			Time:       now,
+			Value:      m.GetCounter().GetValue(),
+			Exemplars:  []metricdata.Exemplar[float64]{convertExemplar(m.GetCounter().GetExemplar())},
+		}
+		createdTs := m.GetCounter().GetCreatedTimestamp()
+		if createdTs.IsValid() {
+			dp.StartTime = createdTs.AsTime()
+		}
+		if m.GetTimestampMs() != 0 {
+			dp.Time = time.UnixMilli(m.GetTimestampMs())
+		}
+		otelCounter.DataPoints[i] = dp
+	}
+	return otelCounter
+}
+
+func convertHistogram(metrics []*dto.Metric, now time.Time) metricdata.Histogram[float64] {
```

Comment:
I think this does not support Native Histograms (exponential ones): https://github.com/prometheus/client_model/blob/master/io/prometheus/client/metrics.proto#L82

That's fine, but let's provide clean error/warning and TODO (: